### PR TITLE
MLDB-2139 plugin dependencies

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: server/mldb_server.cc [hunk 1]
```diff
@@ -404,47 +404,68 @@ scanPlugins(const std::string & dir_)
...
+            logger->error() << "details:";
+            logger->error() << jsonEncode(exc.details);
+            logger->error() << "plugin will be ignored";
+            return;
+        } catch (const std::exception & exc) {
+            logger->error() << "loading plugin " << dir << ": " << exc.what();
+            logger->error() << "plugin will be ignored";
+            return;
+        }
+    };
```
Comment (FinchPowers) [inline server/mldb_server.cc]: This error handler was copy pasted. I extracted it.

File: server/mldb_server.cc [hunk 2]
```diff
@@ -404,47 +404,68 @@ scanPlugins(const std::string & dir_)
...
+
+    auto loadPlugin = [&] (PluginManifest & manifest) {
```
Comment (FinchPowers) [inline server/mldb_server.cc]: Same code as before with the error handler extracted.

File: server/mldb_server.cc [hunk 3]
```diff
@@ -404,47 +404,68 @@ scanPlugins(const std::string & dir_)
...
-                    auto config = manifest.config.params.convert<PluginResource>();
-                    config.address = dir;
-                    manifest.config.params = config;
-                    auto plugin = plugins->obtainEntitySync(
-                        manifest.config, nullptr /* on progress */);
+                for (const auto & dep: manifest.dependencies) {
+                    if (loadedPlugins.find(dep) == loadedPlugins.end()) {
+                        missingDependencies[manifest.config.id] = manifest;
+                        return;
+                    }
```
Comment (FinchPowers) [inline server/mldb_server.cc]: Don't load the plugin now if there are missing deps.

File: server/mldb_server.cc [hunk 4]
```diff
@@ -474,20 +495,36 @@ scanPlugins(const std::string & dir_)
...
-            return;
-        } catch (const std::exception & exc) {
-            logger->error() << "error scanning plugin directory  "
-                            << dir << ": " << exc.what();
-            logger->error() << "plugins will be ignored";
-            return;
+        });
+    }
+
+    // Dumb algo, loop until nothing is left or the last pass didn't load anything
```
Comment (FinchPowers) [inline server/mldb_server.cc]: We resolve the dependencies with a dumb loop. There should not be a lot so perf is not an issue.

**Comment by FinchPowers** [comment]
If someone has any ideas on how to test this, I'm opened to suggestions.

**Comment by FinchPowers** [comment]
Major path bug found. Will fix.

File: server/plugin_manifest.h [hunk 5]
```diff
@@ -21,6 +21,7 @@ namespace MLDB {
 /*****************************************************************************/
 
 struct PluginManifest {
+    std::vector<std::string> dependencies;
```
Comment (jeremybarnes) [inline server/plugin_manifest.h]: Should be UTF8String for consistency probably.

**Comment by jeremybarnes** [comment]
My feeling is that this would be much cleaner if we simply added a "require" function for the dependencies, that was called when loading.  For example, in the main plugin script we would write

mldb.require('tensorflow')

which would then load up the given plugin if it wasn't already loaded.  It keeps it declarative, and allows us to avoid having huge lists of hardcoded dependencies which brings in the kitchen sink every time.

In the redesign of the Python plugin system, we could do something similar: eg mldb.pythonRequires('pippackagename >0.34')

**Pull request closed without merging**